### ISSUE TITLE
VAE does not learn anything

### ISSUE THREAD

**Issue by roman-kh** [blei-lab/edward#419]
I implemented [VAE using Edward](https://github.com/roman-kh/pynotebooks/blob/master/edwardlib/MNIST_vae_edward.ipynb). 
However, it hardly learns anything usefull regardless of the number of epochs.

I also tried to run [your example](https://github.com/blei-lab/edward/blob/master/examples/vae.py). But the result is the same.

Am I doing something terribly wrong?

**Comment by dustinvtran** [comment]
my guess is it is learning but needs some hyper parameter tweaking. 

you can look at the model's hidden representation, that is, before the bernoulli sampling. this will make it a bit more transparent if the model parameters can produce digits. to do this, change the line
```python
imgs = x_decoded.value().eval()
```
to
```python
imgs = dlogits.eval()
```
these are the "blurry" images you typically see in VAE papers. looking at the bernoulli draws is even more difficult to produce effectively, which is why most papers don't show them.

as for the included example in edward, you should see something like the below after only a few epochs.
![screen shot 2017-01-24 at 10 51 24 am](https://cloud.githubusercontent.com/assets/2569867/22255200/9729d6c4-e224-11e6-8621-bc785abab0e8.png)
you can also contrast this with the hidden representation version, which is also after only a few epochs.
![screen shot 2017-01-24 at 10 53 20 am](https://cloud.githubusercontent.com/assets/2569867/22255205/a12d0646-e224-11e6-8f75-b71a99174fb7.png)

as another tip, i highly recommend you keep tracking of the loss function over epochs as you do the training. this is one diagnostic for tracking convergence rather than running for a fixed # of iterations.

**Comment by dustinvtran** [comment]
closing as i think above solves it.

**Comment by roman-kh** [comment]
Eventually I figured out that in order to get images of the digits one should look at probabilities (not bernoulli samples or logits)
`probs = tf.nn.sigmoid(dlogits)`
`imgs = probs.eval()`
